### ISSUE TITLE
Problems with models loaded using Assimp

### ISSUE THREAD

**Issue by Epihaius** [panda3d/panda3d#366]
With the latest Panda development builds for Windows it is possible to load model formats that are supported by Assimp, such as the .obj format.
However, it seems that .obj, along with other formats such as .flt, .lwo, .dxf, .wrl and .x could already be loaded without Assimp support in Panda.

The results of loading an .obj file with and without Assimp support are quite different, though. What they do have in common, is that the loaded models are incorrectly oriented, as if they had a pitch value of -90.

Much worse however, is that .obj models loaded through Assimp appear to be turned inside-out and lit incorrectly. It seems the normals are correct, but the winding order of the vertices is reversed.
Furthermore, a plain PandaNode gets created between the ModelRoot and the GeomNodes for some reason. This PandaNode has the same name as the ModelRoot.

EDIT:

After testing more file formats (.x, .3ds, .fbx, .dxf, .ply, .raw and .stl), it would appear that models loaded from any of these formats also suffer from incorrect face culling, while only 3 of these formats (.ply, .raw and .stl) do not lead to incorrectly oriented models. Interestingly, the latter 3 are not parented to an intermediate PandaNode (which is perhaps used to convert between coordinate systems?).

Models loaded from .ply have a CullFaceAttrib applied that hides the fact that their faces are not correctly culled by default.

It was not possible to test the .ms3d (Milkshape) format, as I got this error upon loading:
`:assimp(error): MS3D: Failure reading comment, length field is out of range`

The following code snippet should reproduce these issues:

```
from panda3d.core import *
from direct.showbase.ShowBase import ShowBase

loadPrcFileData("", "load-file-type p3assimp")


class MyApp(ShowBase):

    def __init__(self):

        ShowBase.__init__(self)

        dir_light = DirectionalLight("light")
        light_np = self.camera.attach_new_node(dir_light)
        light_np.set_hpr(-20., -20., 0.)
        self.render.set_light(light_np)

        model = self.loader.load_model(path_to_model_file, noCache=True)
        model.reparent_to(self.render)
        model.ls()

        # comment out the following line to fix the problems with the loaded model
        return

        # correct the orientation of the model (not needed for .ply, .raw or .stl)
        model.set_p(90.)

        # reverse the winding order of the GeomPrimitives' vertices
        for geom_node in model.find_all_matches("**/+GeomNode"):
            for i in range(geom_node.node().get_num_geoms()):
                geom_node.node().modify_geom(i).reverse_in_place()


app = MyApp()
app.run()
```

Here is a comparison between an .obj model loaded with and without Assimp support:

With Assimp support:
* wrong orientation;
* winding order of GeomPrimitive vertices is reversed;
* GeomNodes are parented to additional (unnecessary?) plain PandaNode under ModelRoot;
* different model objects in the .obj file are created as different GeomNodes;
* materials are loaded.

Without Assimp support:
* wrong orientation;
* winding order of GeomPrimitive vertices is preserved;
* GeomNodes are parented to plain PandaNode; no ModelRoot is created;
* different model objects in the .obj file are created as a single GeomNode;
* materials are ignored.

Tested with Panda build 71e18eb, Python 3.6 64-bit on Windows 10.

**Comment by bluekyu** [comment]
> winding order of GeomPrimitive vertices is reversed;

This can be resolved by removing `aiProcess_FlipWindingOrder` in the below code:
https://github.com/panda3d/panda3d/blob/f12bc29d6d4b28124a6db9c8ba08e963277ed5b6/pandatool/src/assimp/assimpLoader.cxx#L108

It may work, but I did not test it in Panda3D (Python + internal assimp).
(I only checked it works in Panda3D (C++) with customized Assimp loader.)

**Comment by rdb** [comment]
Would it be too much to ask for you to gather some models that show these issues and post them here?

**Comment by Epihaius** [comment]
Of course!

Here is a .zip file containing the Blender monkey that I myself exported to .3ds, .blend, .dxf, .fbx, .obj, .ply, .raw, .stl and .x:

[monkey.zip](https://github.com/panda3d/panda3d/files/2260426/monkey.zip)


And here are links to some free 3D models from free3d.com that I tested:

https://free3d.com/3d-model/low-poly-tree-73217.html (19.3 KB; .3ds, .fbx, .obj)
https://free3d.com/3d-model/usp-45-44387.html (188 KB; .3ds, .blend, .fbx, .obj)
https://free3d.com/3d-model/low-poly-deer-72513.html (300 KB; .3ds, .c4d, .dae, .fbx, .obj)
https://free3d.com/3d-model/low-poly-cat-46138.html (430 KB; .3ds, .c4d, .dae, .fbx, .obj)
https://free3d.com/3d-model/eyeball--33237.html (3.9 MB; .3ds, .blend, .dxf, .dae, .fbx, .obj, .ply, .stl, .x3d)
https://free3d.com/3d-model/monitor-79124.html (6 MB; .abc, .blend, .dae, .fbx, .obj, .stl, .x3d)
https://free3d.com/3d-model/notebook-low-poly-version-57341.html (10.8 MB; .3ds, .blend, .dae, .dxf, .fbx, .obj, .x, .x3d)
I'd give you some fries with that, but that particular model wasn't free ;) .

All versions of each of these models are rendered inside-out when loaded.
It would seem that loaded .blend files are oriented correctly, just like the .ply, .raw and .stl formats.
Models loaded from all other file formats have a pitch of -90 degrees, for most formats explicitly set on a parent PandaNode, but not all (in the case of .obj, this transform seems baked into the model).

**Comment by rdb** [comment]
Thanks!  That's, well, damning evidence.

I changed it to not flip the winding order, and made it configurable.  In fact, I made a lot of flags configurable.  These are the Config.prc settings I added, with their respective defaults:
```

# Calculates tangents and binormals for meshes imported via Assimp.
assimp-calc-tangent-space false


# Merges duplicate vertices.  Set this to false if you want each
# vertex to only be in use on one triangle.
assimp-join-identical-vertices true


# Improves rendering performance of the loaded meshes by reordering
# triangles for better vertex cache locality.  Set this to false if
# you need geometry to be loaded in the exact order that it was
# specified in the file, or to improve load performance.
assimp-improve-cache-locality true


# Removes redundant/unreferenced materials from assets.
assimp-remove-redundant-materials true


# Determines which normal vectors are facing inward and inverts them
# so that they are facing outward.
assimp-fix-infacing-normals false


# Removes the number of draw calls by unifying geometry with the same
# materials.  Especially effective in conjunction with
# assimp-optimize-graph and assimp-remove-redundant-materials.
assimp-optimize-meshes true


# Optimizes the scene geometry by flattening the scene hierarchy.
# This is very efficient (combined with assimp-optimize-meshes), but
# it may result the hierarchy to become lost, so it is disabled by
# default.
assimp-optimize-graph false


# Set this true to flip the winding order of all models loaded via
# the Assimp loader.  Note that you may need to clear the model-cache
# after changing this.
assimp-flip-winding-order false
```

Note that you need to clear the model-cache after changing any of these variables.  But if you are using this in an open/import dialog in Panda3D Studio, then you probably are loading with `noCache=True` already anyway.

**Comment by rdb** [comment]
It seems that Assimp does not have a notion of which coordinate system the models are in, so I don't think we can do much about the fact that some models are loaded rotated.

Is there anything else I missed or can this issue be closed?

**Comment by Epihaius** [comment]
Wow, that's a lot of configuration options, awesome!

Concerning the rotated models, I really do feel that by default their pitch should be offset by 90 degrees, so the models loaded from the majority of the formats (which actually includes the most popular formats IMHO) will have the correct orientation. That would however be the wrong choice for the .blend format, so a lot of users might still complain then, hmmm. Could it perhaps be made into another config option, by default set to true?

There is one last format I didn't mention yet, but which was already reported to be problematic in [this forum post](https://discourse.panda3d.org/t/panda3d-p3assimp-in-windows/23854/10): the .gltf format (and the associated .glb format).
Trying to load a .glb file exported from Blender failed with the following error:

`:assimp(error): GLTF: Unable to read the file header`

while loading any .gltf file leads to an empty model (a ModelRoot node with nothing attached to it).

To test the .gltf issue you could download [these models](https://github.com/flemos/GLTF-Free).
Unfortunately I was unable to find any freely downloadable .glb files (not without an account, anyway).

**Comment by rdb** [comment]
Unfortunately, these formats do not actually encode the coordinate system, and even when you export from Blender to these formats you have a choice to export them in whatever coordinate system you'd like.  So if people try to do the right thing and enter Z-up, hoping to match Panda's coordinate system, then they would really want it to be loaded as-is without Panda applying an additional rotation.

Adding a configuration option seems to make some sense, but then it should really only apply to the formats that are coordinate-system-agnostic and do not specify or prescribe a particular coordinate system.  For example, a glTF file is specified to always be Y-up, so if you find that a glTF file is not being loaded correctly, that is something we should definitely fix.

As for glb, perhaps you are loading a glTF 2.0 file whereas Assimp only supports glTF 1.0, or vice-versa?  The Khronos Group provides a large repository of various glTF 1.0 and 2.0 sample models to test with:
https://github.com/KhronosGroup/glTF-Sample-Models

For what it's worth, we hope to soon have a native glTF 2.0 loader in Panda3D.

**Comment by Epihaius** [comment]
Understood about the coordinate system issue; best to leave it up to the user then.

Thanks for the link to the glTF sample models.

Both glTF 1.0 and glTF 2.0 files seem to cause problems for the Assimp loader. This is strange, as the [latest release](https://github.com/assimp/assimp/releases) (4.1.0, from Dec 11, 2017) clearly states its support of glTF 2 import/export. Then again, I don't know if Panda uses that release, or perhaps an older one?

Anyway, when I try to load e.g. the version 1.0 Box.gltf, I get this error:

`:assimp(error): GLTF: error while reading referenced file "Box.bin"`

(even though the .bin file is in the same directory as the .gltf file),
while the version 2.0 Box.gltf results in an empty model again.

Trying to load either the version 1.0 Box.glb or version 2.0 Box.glb leads to this error:
`:assimp(error): GLTF: Unable to read the file header`
in both cases.

I've tried other models as well, with the same results. So there really must be things going wrong with these formats.
But since these might not be Panda-related (and seeing there could soon be other solutions), I'm going to close this issue now.

**Comment by Moguri** [comment]
As a heads up, [this project](https://github.com/Moguri/panda3d-gltf) can be used to import glTF (no binary glTF support yet) into Panda.

**Comment by rdb** [comment]
The problems related to loading glTF files through Assimp have been fixed on the 1.10 branch, although it is still recommended to use panda3d-gltf instead.  I have opened #1386 to address the coordinate system issue.

The extra root node being created is a quirk of Assimp's .obj importer:
https://github.com/assimp/assimp/blob/9a5d6282322de3085474ca8f9f4a4a8f84756394/code/AssetLib/Obj/ObjFileImporter.cpp#L164-L172

It appears that many importers have this quirk.  I am adding an `assimp-collapse-dummy-root-node` variable that can be set to `true` to detect and remove this root node based on some heuristics. It will be set to `true` by default in Panda3D 1.11.0.
